src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -0,0 +1,171 @@
+        return (
+            m for m in cls._get_all_abstract_methods() if getattr(m, _IS_RPC_METHOD_ATTR, False)
+        )
+
+    @classmethod
+    def _declares_service_interface(cls) -> bool:
+        """Check whether a subclass declares the service interface.
+
+        By "service interface", we mean the set of RPC methods that are offered.
+        Those methods are decorated by `@rpc_method`, so we return true for any
+        service class that declares (but does not implement) at least one such method.
+
+        This method would return false, for example, for the local, database-backed
+        implementation that inherits from the base service. It also would return
+        false on an intermediate, abstract subclass that is meant to be extended to
+        declare other base services.
+        """
+
+        try:
+            next(cls._get_all_abstract_rpc_methods())
+        except StopIteration:
+            return False
+        else:
+            return True
+
+    @classmethod
+    @abstractmethod
+    def get_local_implementation(cls) -> "RpcService":
+        """Return a service object that runs locally.
+
+        The returned service object is (generally) the database-backed instance that
+        is called when we either receive a remote call from outside, or want to call
+        it within the same silo.
+
+        A base service class generally should override this class method, making a
+        forward reference to its own database-backed subclass.
+        """
+
+        raise NotImplementedError
+
+    @classmethod
+    def _create_local_delegation(cls) -> "RpcService":
+        """Create a lazy wrapper around the local implementation.
+
+        The indirection is necessary so that a base service's
+        `get_local_implementation` body can safely make a forward reference that
+        `resolve_to_delegation` won't call yet.
+        """
+        impl_attrname = "__cached_local_implementation"
+
+        @cached_property  # type: ignore
+        def cached_local_implementation(service_obj: "RpcService") -> "RpcService":
+            return service_obj.get_local_implementation()
+
+        def create_delegating_method(base_method: Callable[..., Any]) -> Callable[..., Any]:
+            method_name = base_method.__name__
+
+            def delegating_method(service_obj: "RpcService", *args: Any, **kwargs: Any) -> Any:
+                impl = getattr(service_obj, impl_attrname)
+                method = getattr(impl, method_name)
+                return method(*args, **kwargs)
+
+            return delegating_method
+
+        overrides = {
+            service_method.__name__: create_delegating_method(service_method)
+            for service_method in cls._get_all_abstract_methods()
+        }
+        overrides[impl_attrname] = cached_local_implementation
+        remote_service_class = type(f"{cls.__name__}__LocalDelegate", (cls,), overrides)
+        return cast(RpcService, remote_service_class())
```

Comment:
```suggestion
        local_service_class = type(f"{cls.__name__}__LocalDelegate", (cls,), overrides)
        return cast(RpcService, local_service_class())
```

Using local here would align terminology more.